web/gds-admin-ui/src/helpers/api/apiCore.js

```diff
@@ -186,11 +199,29 @@ class APICore {
 
         const decodedAccessToken = jwtDecode(user.access_token);
         const currentTime = Date.now() / 1000;
-        if (decodedAccessToken.exp < currentTime) {
+        const csrfToken = getCookie('csrf_token');
+
+        if (currentTime < decodedAccessToken.exp && currentTime > decodedAccessToken.nbf) {
+            if (isValidRefreshToken(user.refresh_token)) {
+
+                const payload = {
+                    credential: user.refresh_token
+                }
+                axios.post('/reauthenticate', payload, {
+                    headers: {
+                        'X-CSRF-TOKEN': csrfToken
+                    }
+                }).then(res => {
+                    this.setLoggedInUser(res.data)
+                    setAuthorization(res.data.access_token)
+                }).catch((err) => {
+                    console.log('[/reauthenticate] error----', err)
+                })
+            }
+            return true;
+        } else {
```

Comment:
This logic is not quite right:

```javascript
if (currentTime < decodedAccessToken.exp && currentTime > decodedAccessToken.nbf) {
    // The access token is valid -- we could just return true here
    // Alternatively, we could check if we're in that small window of time where we can reauthenticate when the access token is valid:
    if (isValidRefreshToken(user.refresh_token)) {
      const rep = await reauthenticate();
      return rep;
    }
    return true;
} else {
    // access token is invalid, check if we can reauthenticate
    if (isValidRefreshToken(user.refresh_token)) {
      const rep = await reauthenticate();
      return rep;
    }
    // neither the access nor the refresh token is valid any longer
   return false;
}
```

Reply:
There concern is that I verified both tokens 
- The refresh token`nbf` is `35min` after token creation
- The refresh token expire at the same time with `access token` that is 1hour
Then I was expecting that the access token expire before the refresh token to be able to make a `reauthenticate` request if there is an `error`